src/Core/MySQL/PacketsProtocolBinary.cpp

```diff
@@ -0,0 +1,363 @@
+
+                        static constexpr UInt32 MaxScale = DecimalUtils::max_precision<DateTime64>;
+                        scale = scale > MaxScale ? MaxScale : scale;
+
+                        const auto dt64 = assert_cast<const ColumnDateTime64 &>(*col).getData()[row_num];
+                        auto components = DecimalUtils::split(dt64, scale);
+
+                        using T = typename DateTime64::NativeType;
+                        if (dt64.value < 0 && components.fractional)
+                        {
+                            components.fractional
+                                = DecimalUtils::scaleMultiplier<T>(scale) + (components.whole ? T(-1) : T(1)) * components.fractional;
+                            --components.whole;
+                        }
+                        if (scale > 6)
+                        {
+                            // MySQL Timestamp has max scale of 6
+                            components.fractional /= static_cast<int>(pow(10, scale - 6));
+                        }
+
+                        LocalDateTime ldt = LocalDateTime(components.whole, DateLUT::instance(getDateTimeTimezone(*data_type)));
+
+                        bool has_microseconds = components.fractional != 0;
+                        bool has_time = !(ldt.hour() == 0 && ldt.minute() == 0 && ldt.second() == 0);
+                        if (has_microseconds)
+                        {
+                            payload_size += 12;
+                        }
+                        else if (has_time)
+                        {
+                            payload_size += 8;
+                        }
+                        else
+                        {
+                            payload_size += 5;
+                        }
+                        break;
+                    }
+                    default:
+                        WriteBufferFromOwnString ostr;
+                        serializations[i]->serializeText(*columns[i], row_num, ostr, format_settings);
+                        payload_size += getLengthEncodedStringSize(ostr.str());
+                        serialized[i] = std::move(ostr.str());
+                        break;
+                }
+            }
+        }
+
+        void ResultSetRow::writePayloadImpl(WriteBuffer & buffer) const
+        {
+            buffer.write(static_cast<char>(0x00));
+            buffer.write(null_bitmap.data(), null_bitmap_size);
+            for (size_t i = 0; i < columns.size(); ++i)
+            {
+                ColumnPtr col = columns[i];
+                if (col->isNullAt(row_num))
+                {
+                    continue;
+                }
+
+                DataTypePtr data_type = removeLowCardinality(removeNullable((data_types[i])));
+                TypeIndex type_index = data_type->getTypeId();
+                switch (type_index)
+                {
+                    case TypeIndex::UInt8: {
```

Comment:
Could we retain the same order of cases in this method as in the constructor? UInt8, Int8, UInt16, Int16, ...